backend/cmd/update/main.go

```diff
@@ -0,0 +1,363 @@
+	update.triggerUpdate = flag.Bool("upgrade", false, "Upgrade molly wallet binary")
+	flag.Parse()
+
+	// if trigger update, update molly
+	// if errors trigger RestoreBackup
+
+	fmt.Println(*update.dagFolderPath, *update.triggerUpdate, *update.newVersion, *update.oldMollyBinaryPath)
+	update.Run()
+
+	//fmt.Printf("Dag Folder: %s, Current Version: %s, Molly Path: %s, New Version: %s, Update: %v\n", *update.dagFolderPath, *update.currentVersion, *update.oldMollyBinaryPath, *update.newVersion, *update.triggerUpdate)
+}
+
+func (u *Update) Run() {
+	var err error
+
+	// Clean up old update artifacts
+	err = u.CleanUp()
+	if err != nil {
+		log.Fatalf("Unable to clear previous local state: %v", err)
+	}
+
+	// Create a TCP connection to localhost on port 36866
+	u.clientRPC, err = rpc.DialHTTP("tcp", "localhost:36866")
+	if err != nil {
+		log.Fatal("Connection error: ", err)
+	}
+	log.Infof("Successfully established RPC connection with Molly Wallet")
+	defer u.clientRPC.Close()
+
+	zippedArchive, err := u.DownloadAppBinary()
+	if err != nil {
+		log.Fatalf("Unable to download v%s of Molly Wallet: %v", *u.newVersion, err)
+	}
+
+	// TODO: checksum verification
+
+	contents, err := unzipArchive(zippedArchive, *u.dagFolderPath)
+	if err != nil {
+		log.Fatalf("Unable to unzip contents: %v", err)
+	}
+
+	err = u.BackupApp()
+	if err != nil {
+		log.Fatalf("Unable to Backup Molly Wallet: %v", err)
+	}
+
+	err = u.TerminateAppService()
+	if err != nil {
+		log.Fatalf("Unable to terminate Molly Wallet: %v", err)
+		err = u.RestoreBackup()
+		if err != nil {
+			log.Fatal("Unable to restore backup.")
+		}
+	}
+
+	err = u.ReplaceAppBinary(contents)
+	if err != nil {
+		log.Fatalf("Unable to overwrite old installation: %v", err)
+		err = u.RestoreBackup()
+		if err != nil {
+			log.Fatal("Unable to restore backup.")
+		}
+	}
+
+	err = u.LaunchAppBinary()
+	if err != nil {
+		log.Fatalf("Unable to start up Molly after update: %v", err)
+		err = u.RestoreBackup()
+		if err != nil {
+			log.Fatal("Unable to restore backup.")
+		}
+	}
+
+	err = u.CleanUp()
+	if err != nil {
+		log.Fatalf("Unable to clear previous local state: %v", err)
+	}
+
+}
+
+// DownloadAppBinary downloads the latest Molly Wallet zip from github releases and returns the path to it
+func (u *Update) DownloadAppBinary() (string, error) {
+
+	filename := "mollywallet.zip"
+	osBuild, _ := getUserOS() // returns linux, windows, darwin or unsupported as well as the file extension (e.g ".exe")
+
+	if osBuild == "unsupported" {
+		return "", fmt.Errorf("The OS is not supported.")
+	}
+
+	url := u.downloadURL + "/v" + *u.newVersion + "-" + osBuild + "/" + filename
+	// e.g https://github.com/grvlle/constellation_wallet/releases/download/v1.1.9-linux/mollywallet.zip
+	log.Infof("Constructed the following URL: %s", url)
+
+	out, err := os.Create(*u.dagFolderPath + "/" + filename + ".tmp")
+	if err != nil {
+		return "", err
+	}
```

Comment:
```suggestion
	filePath := path.Join(*u.dagFolderPath, filename)
	tmpFilePath := filePath + ".tmp"
	out, err := os.Create(tmpFilepath)
	if err != nil {
		return "", err
	}
	defer out.Close()
```

Use `defer` instead of calling `out.Close()` in each path below.